aten/src/ATen/native/cuda/FractionalMaxPool2d.cu

```diff
@@ -0,0 +1,375 @@
+  // Output (h, w) point that this thread is responsible for
+  int ourOutputPoint = threadIdx.x + blockIdx.x * blockDim.x;
+  int plane = blockIdx.y;
+  int batch = blockIdx.z;
+
+  // Each thread generates a specific output point
+  if (ourOutputPoint < gradOutput.size(2) *
+    gradOutput.size(3)) {
+    int outputW = ourOutputPoint % gradOutput.size(3);
+    int outputH = ourOutputPoint / gradOutput.size(3);
+
+    int index = indices[batch][plane][outputH][outputW];
+    assert(index >= 0);
+    int inputW = index % gradInput.size(3);
+    int inputH = index / gradInput.size(3);
+    assert(inputH < gradInput.size(2));
+
+    atomicAdd(
+      &gradInput[batch][plane][inputH][inputW],
+      gradOutput[batch][plane][outputH][outputW]
+    );
+  }
+}
+
+void fractional_max_pool2d_out_cuda_template(
+  Tensor & output,
+  Tensor& indices,
+  const Tensor& input,
+  IntList pool_size,
+  IntList output_size,
+  const Tensor& randomSamples) {
+  int planeDim = 0;
+  int dimh = 1;
+  int dimw = 2;
+  int numBatch = 1;
+
+  int ndims = input.ndimension();
+  for (int i = 0; i < ndims; i++) {
+     AT_CHECK(input.size(i) > 0,
+       "fractional_max_pool2d(): expected input to have non-empty spatial dimensions, "
+       "but input has sizes ", input.sizes(), " with dimension ", i, " being "
+       "empty");
+   }
+
+   AT_CHECK((ndims == 3 || ndims == 4),
+     "non-empty 3D or 4D (batch mode) tensor expected for input");
+
+  if (ndims == 4) {
+    numBatch = input.size(0);
+    planeDim++;
+    dimh++;
+    dimw++;
+  }
+
+  /* sizes */
+  int numPlanes = input.size(planeDim);
+  int inputH = input.size(dimh);
+  int inputW = input.size(dimw);
+
+  int outputH = output_size[0];
+  int outputW = output_size[1];
+  int poolSizeH = pool_size[0];
+  int poolSizeW = pool_size[1];
+
+  AT_CHECK(outputH + poolSizeH - 1 <= inputH,
+             "fractional_max_pool2d(): pool_size height ", poolSizeH,
+             " too large relative to input height ", inputH);
+  AT_CHECK(outputW + poolSizeW - 1 <= inputW,
+           "pool_size width ", poolSizeW,
+           " too large relative to input width ", inputW);
+
+  if (ndims == 3) {
+    /* resize output */
+    output.resize_({numPlanes, outputH, outputW});
+    /* indices will contain the locations for each output point */
+    indices.resize_({numPlanes, outputH, outputW});
+  } else {
+    output.resize_({numBatch, numPlanes, outputH, outputW});
+    indices.resize_({numBatch, numPlanes, outputH, outputW});
+  }
+
+  auto output_ = output;
+  auto input_ = input;
+  auto indices_ = indices;
+
+  if(ndims == 3) {
+    output_.reshape({1, numPlanes, outputH, outputW});
+    indices_.reshape({1, numPlanes, outputH, outputW});
+    input_.reshape({1, input.size(0), input.size(1), input.size(2)});
+  }
+
+  // block is limited to 4 warps
+  // grid handles overflow per each plane
+  int outputPlaneSize = output_.size(2) *
+    output_.size(3);
+  dim3 grid((outputPlaneSize + 127) / 128,
```

Comment:
nit: Any particular reason you inlined THCCeilDiv here? The inlining makes the intent of the line here less clear

Reply:
I wanted to remove all header files from THNN/THCUNN, so I moved the inline implementation here. I also thought it was quite common in CUDA (it appeared in the elementary CUDA tutorial: https://devblogs.nvidia.com/even-easier-introduction-cuda/).

I am going to add a comment line for readability; let me know if you suggest otherwise.